<a href="https://colab.research.google.com/github/MatveiV/ML_Fin_Notebooks/blob/main/%D0%A2%D0%B5%D0%BC%D0%B0_11_4_%D0%A3%D1%85%D1%83%D0%B4%D1%88%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B0%D0%B2%D1%82%D0%BE%D0%B2%D1%8B%D0%B1%D0%BE%D1%80%D0%B0_%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D0%B8_%D0%BD%D0%B0_%D0%BD%D0%B5%D0%BE%D1%87%D0%B8%D1%89%D0%B5%D0%BD%D0%BD%D1%8B%D1%85_%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Автоматический подбор моделей машинного обучения

Автоматическое машинное обучение не заменяет специалиста по данным, но оно может помочь вам быстрее находить хорошие модели. Одним из фреймворков является [TPOT](https://github.com/EpistasisLab/tpot) (Tree-based Pipeline Optimization Tool), который позиционирует себя как Ваш помощник по науке о данных.

TPOT должен быть помощником, который дает вам идеи о том, как решить конкретную проблему машинного обучения, исследуя конфигурации конвейера, которые вы, возможно, никогда не рассматривали, а затем оставляет тонкую настройку более ограниченным методам настройки параметров, таким как поиск по сетке.

Таким образом, TPOT помогает вам находить хорошие алгоритмы.

In [ ]:
!pip install tpot -qq

In [ ]:
from tpot import TPOTClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

digits = load_iris()
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.55, test_size=0.45, random_state=42)

tpot = TPOTClassifier(generations=5, population_size=50, verbosity=2, random_state=42)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))

Optimization Progress:   0%|          | 0/300 [00:00<?, ?pipeline/s]


Generation 1 - Current best internal CV score: 0.9647058823529411

Generation 2 - Current best internal CV score: 0.9647058823529411

Generation 3 - Current best internal CV score: 0.9647058823529411

Generation 4 - Current best internal CV score: 0.9647058823529411

Generation 5 - Current best internal CV score: 0.9647058823529411

Best pipeline: MLPClassifier(input_matrix, alpha=0.0001, learning_rate_init=0.001)
1.0


In [ ]:
import random
for i in X_train:
  for j in range(len(i)):
    X_train[random.randint(0,len(X_train)-1)][random.randint(0,len(i)-1)]=None
X_train

array([[4.6, 3.2, nan, 0.2],
       [6.2, 2.9, nan, nan],
       [nan, 2.5, nan, nan],
       [5.5, nan, nan, nan],
       [nan, nan, 4.8, nan],
       [nan, nan, nan, 1.9],
       [6. , 2.2, 4. , nan],
       [nan, nan, 4.5, nan],
       [6.2, nan, nan, 2.3],
       [nan, nan, nan, 1.3],
       [nan, nan, nan, nan],
       [5. , nan, nan, nan],
       [6.4, nan, nan, nan],
       [nan, nan, 1.4, 0.2],
       [5. , 3.2, nan, nan],
       [nan, 2.4, 3.8, nan],
       [nan, nan, nan, 1.7],
       [nan, 3.1, nan, nan],
       [nan, nan, 5.1, nan],
       [nan, nan, nan, 0.2],
       [nan, nan, 1.6, nan],
       [5.9, 3.2, nan, nan],
       [nan, nan, nan, nan],
       [nan, 3.2, nan, nan],
       [6. , 2.7, 5.1, nan],
       [nan, nan, nan, 1.4],
       [7.7, nan, nan, nan],
       [nan, nan, nan, nan],
       [4.4, nan, nan, nan],
       [nan, 3. , nan, 0.1],
       [nan, 2.2, nan, 1.5],
       [nan, nan, nan, nan],
       [nan, 3.1, 1.5, 0.2],
       [nan, nan, nan, 0.3],
       [nan, n

In [ ]:
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))

Imputing missing values in feature set


Optimization Progress:   0%|          | 0/300 [00:00<?, ?pipeline/s]


Generation 1 - Current best internal CV score: 0.6727941176470589

Generation 2 - Current best internal CV score: 0.6823529411764706

Generation 3 - Current best internal CV score: 0.6838235294117647

Generation 4 - Current best internal CV score: 0.6955882352941176

Generation 5 - Current best internal CV score: 0.6955882352941176

Best pipeline: LogisticRegression(MinMaxScaler(input_matrix), C=25.0, dual=False, penalty=l2)
Imputing missing values in feature set
0.8676470588235294


Вывод: используя один набор данных автоматически был выбран классификатор MLPClassifier с точностью 0.96, на неочищенный набор данных автоматически был выбран классификатор LogisticRegression с точностью 0,69  